### PR TITLE
[c++] Fix display of Arrow schema for enum of `bytes` datatype

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:** Details on #2306 

Checker from #2306:

```
import tiledbsoma as soma
sdf = soma.open('test_dataframe')
print(sdf.schema)
```
 
Output after this PR:

```
soma_joinid: int64 not null
int_cat: dictionary<values=int64, indices=int8, ordered=0> not null
int: int64 not null
str_cat: dictionary<values=string, indices=int8, ordered=0> not null
str: large_string not null
byte_cat: dictionary<values=binary, indices=int8, ordered=0> not null
byte: large_binary not null
```

**Changes:** Report binary appropriately

**Notes for Reviewer:** Failing CI

### CODE REVIEW

**Comment by johnkerl** [comment]
One problem: when I re-run @bkmartinjr 's data-generator script (the one in the description of this PR which creates `test_dataframe`) now its output is like this:

```
$ ./2305.py
** Original Pandas schema
soma_joinid       int64
int_cat        category
int               int64
str_cat        category
str              object
byte_cat       category
byte             object
dtype: object
soma_joinid: dtype('int64')
int_cat: CategoricalDtype(categories=[10, 20], ordered=False, categories_dtype=int64)
int: dtype('int64')
str_cat: CategoricalDtype(categories=['A', 'B'], ordered=False, categories_dtype=object)
str: dtype('O')
byte_cat: CategoricalDtype(categories=[b'A', b'B'], ordered=False, categories_dtype=object)
byte: dtype('O')
-----
** Arrow schema, derived from Pandas
soma_joinid: int64
int_cat: dictionary<values=int64, indices=int8, ordered=0>
int: int64
str_cat: dictionary<values=string, indices=int8, ordered=0>
str: string
byte_cat: dictionary<values=binary, indices=int8, ordered=0>
byte: binary
-- schema metadata --
pandas: '{"index_columns": [], "column_indexes": [], "columns": [{"name":' + 964
-----
** Arrow Table derived from pandas
pyarrow.Table
soma_joinid: int64
int_cat: dictionary<values=int64, indices=int8, ordered=0>
int: int64
str_cat: dictionary<values=string, indices=int8, ordered=0>
str: string
byte_cat: dictionary<values=binary, indices=int8, ordered=0>
byte: binary
----
soma_joinid: [[0,1,2,3,4,5]]
int_cat: [  -- dictionary:
[10,20]  -- indices:
[0,1,0,1,1,1]]
int: [[10,20,10,20,20,20]]
str_cat: [  -- dictionary:
["A","B"]  -- indices:
[0,1,0,1,1,1]]
str: [["A","B","A","B","B","B"]]
byte_cat: [  -- dictionary:
[41,42]  -- indices:
[0,1,0,1,1,1]]
byte: [[41,42,41,42,42,42]]
-----
**Created TileDB Array schema
soma_joinid: int64 not null
int_cat: dictionary<values=int64, indices=int8, ordered=0> not null
int: int64 not null
str_cat: dictionary<values=string, indices=int8, ordered=0> not null
str: large_string not null
byte_cat: dictionary<values=binary, indices=int8, ordered=0> not null
byte: large_binary not null
Traceback (most recent call last):
  File "/Users/johnkerl/Desktop/./2305.py", line 64, in <module>
    main()
  File "/Users/johnkerl/Desktop/./2305.py", line 50, in main
    df = soma_dataframe.read().concat().to_pandas()
  File "/Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/src/tiledbsoma/_read_iters.py", line 74, in concat
    return pa.concat_tables(self)
  File "pyarrow/table.pxi", line 5366, in pyarrow.lib.concat_tables
  File "/Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/src/tiledbsoma/_read_iters.py", line 70, in __next__
    return next(self._reader)
  File "/Users/johnkerl/git/single-cell-data/TileDB-SOMA/apis/python/src/tiledbsoma/_read_iters.py", line 464, in _arrow_table_reader
    tbl = sr.read_next()
  File "pyarrow/array.pxi", line 1628, in pyarrow.lib.Array._import_from_c
  File "pyarrow/error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 100, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Expected 3 buffers for imported type binary, ArrowArray struct has 2
```

**Comment by nguyenv** [comment]
https://github.com/single-cell-data/TileDB-SOMA/tree/viviannguyen/array-write-path

I actually already fixed this in the write path refactor now that enumerations getting handled on the C++ side instead of Python. Overall though, I still have 21 failing unit tests for this PR, so I don't think I can get my fix out immediately...

```
(soma-3.11) vivian@mangonada:~/tiledb-bugs$ python 2305.py
** Original Pandas schema
soma_joinid       int64
int_cat        category
int               int64
str_cat        category
str              object
byte_cat       category
byte             object
dtype: object
soma_joinid: dtype('int64')
int_cat: CategoricalDtype(categories=[10, 20], ordered=False, categories_dtype=int64)
int: dtype('int64')
str_cat: CategoricalDtype(categories=['A', 'B'], ordered=False, categories_dtype=object)
str: dtype('O')
byte_cat: CategoricalDtype(categories=[b'A', b'B'], ordered=False, categories_dtype=object)
byte: dtype('O')
-----
** Arrow schema, derived from Pandas
soma_joinid: int64
int_cat: dictionary<values=int64, indices=int8, ordered=0>
int: int64
str_cat: dictionary<values=string, indices=int8, ordered=0>
str: string
byte_cat: dictionary<values=binary, indices=int8, ordered=0>
byte: binary
-- schema metadata --
pandas: '{"index_columns": [], "column_indexes": [], "columns": [{"name":' + 964
-----
** Arrow Table derived from pandas
pyarrow.Table
soma_joinid: int64
int_cat: dictionary<values=int64, indices=int8, ordered=0>
int: int64
str_cat: dictionary<values=string, indices=int8, ordered=0>
str: string
byte_cat: dictionary<values=binary, indices=int8, ordered=0>
byte: binary
----
soma_joinid: [[0,1,2,3,4,5]]
int_cat: [  -- dictionary:
[10,20]  -- indices:
[0,1,0,1,1,1]]
int: [[10,20,10,20,20,20]]
str_cat: [  -- dictionary:
["A","B"]  -- indices:
[0,1,0,1,1,1]]
str: [["A","B","A","B","B","B"]]
byte_cat: [  -- dictionary:
[41,42]  -- indices:
[0,1,0,1,1,1]]
byte: [[41,42,41,42,42,42]]
-----
**Created TileDB Array schema
soma_joinid: int64 not null
int_cat: dictionary<values=int64, indices=int8, ordered=0> not null
int: int64 not null
str_cat: dictionary<values=string, indices=int8, ordered=0> not null
str: large_string not null
byte_cat: dictionary<values=binary, indices=int8, ordered=0> not null
byte: large_binary not null
soma_joinid: dtype('int64'), dtype('int64')
int_cat: CategoricalDtype(categories=[10, 20], ordered=False, categories_dtype=int64), CategoricalDtype(categories=[10, 20], ordered=False, categories_dtype=int64)
Categories dtype: dtype('int64'), dtype('int64')
int: dtype('int64'), dtype('int64')
str_cat: CategoricalDtype(categories=['A', 'B'], ordered=False, categories_dtype=object), CategoricalDtype(categories=['A', 'B'], ordered=False, categories_dtype=object)
Categories dtype: dtype('O'), dtype('O')
str: dtype('O'), dtype('O')
byte_cat: CategoricalDtype(categories=[b'A', b'B'], ordered=False, categories_dtype=object), CategoricalDtype(categories=[b'A', b'B'], ordered=False, categories_dtype=object)
Categories dtype: dtype('O'), dtype('O')
byte: dtype('O'), dtype('O')
   soma_joinid int_cat  int str_cat str byte_cat  byte
0            0      10   10       A   A     b'A'  b'A'
1            1      20   20       B   B     b'B'  b'B'
2            2      10   10       A   A     b'A'  b'A'
3            3      20   20       B   B     b'B'  b'B'
4            4      20   20       B   B     b'B'  b'B'
5            5      20   20       B   B     b'B'  b'B'
```

File: libtiledbsoma/src/utils/arrow_adapter.cc
Comment (johnkerl) [inline libtiledbsoma/src/utils/arrow_adapter.cc:386]: * @nguyenv noted she has a fix (https://github.com/single-cell-data/TileDB-SOMA/pull/2305#issuecomment-2015509453) but that is large and has CI fails
* This PR is my attempt to extract the essentials to unblock 1.9 for issue #2306 
* When I have `"Z"` here then the repro script in the description field of this PR passes. But then other existing unit-test cases fail, since what should read back as `["a","b"]` reads back as `[b"a", b"a"]`
* Yet when I have `"U"` here, then the repro script in the description field of this PR fails -- the reported problem is simply not fixed -- but all existing unit-test cases pass.

I seem to recall that `STRING_ASCII` _along with_ some other flag, metadata something-something, is used to differentiate between string-as-bytes and string-as-string ... @nguyenv / @ihnorton perhaps you recall how TileDB-Py does this (as it successfully does, in the code snippet included in the description field of this PR).

Comment (nguyenv) [reply libtiledbsoma/src/utils/arrow_adapter.cc:386]: https://github.com/single-cell-data/TileDB-SOMA/compare/main...viviannguyen/array-write-path#diff-c66edd736f9a9cb059c6f6e0f4cc6d1cdfa0fd91a2b9220dd2e83c12c2546089R451-R457
https://github.com/single-cell-data/TileDB-SOMA/compare/main...viviannguyen/array-write-path#diff-c66edd736f9a9cb059c6f6e0f4cc6d1cdfa0fd91a2b9220dd2e83c12c2546089R407

I believe these are the snippets that should fix the issue you're seeing.

Comment (nguyenv) [reply libtiledbsoma/src/utils/arrow_adapter.cc:386]: I specially handle the string types for enums like that rather relying on `to_tiledb_format`.

Comment (johnkerl) [reply libtiledbsoma/src/utils/arrow_adapter.cc:386]: @nguyenv both of the links you sent me open up with "57 files changed" -- non-navigable

Nonetheless I'll take a look at what you wrote in Slack 🤞

**Comment by johnkerl** [comment]
Thanks @nguyenv for the advice! CI is green, and my interactive use as noted in the description field of this PR also works great. :D 

cc @ihnorton @jp-dark or anyone else still on line at this late hour for approval; or we can do this Monday.

**Review by ihnorton — APPROVED** [review]
Consider adding the script from https://github.com/single-cell-data/TileDB-SOMA/issues/2306 to a python `repr` test? Otherwise LGTM.

**Comment by johnkerl** [comment]
Thanks @ihnorton 

Delaying the merge for a unit-test case

It'll be more 'somatic' to test it like this:

```
    with soma.open(uri) as sdf:

        f = sdf.schema.field("int_cat")
        assert f.type.index_type == pa.int8()
        assert f.type.value_type == pa.int64()


        f = sdf.schema.field("str_cat")
        assert f.type.index_type == pa.int8()
        assert f.type.value_type == pa.string()


        f = sdf.schema.field("byte_cat")
        assert f.type.index_type == pa.int8()
        assert f.type.value_type == pa.binary() or f.type.value_type == pa.large_binary()
```

with data as ☝️ and I'll get that in before merging

**Comment by johnkerl** [comment]
@nguyenv this is still not right :( 

The original repro script I was using, I had confused myself by putting attrs in the order

```
soma_joinid: int64 not null
int: int64 not null
int_cat: dictionary<values=int64, indices=int8, ordered=0> not null
str: large_string not null
str_cat: dictionary<values=string, indices=int8, ordered=0> not null
byte: large_binary not null
byte_cat: dictionary<values=string, indices=int8, ordered=0> not null
```

A less error-prone ordering is here: https://gist.github.com/johnkerl/6c4a8ee3ceacaaf2e1929873095b2933

And we still have:

```
import tiledbsoma as soma
sdf = soma.open('test_dataframe')
print(sdf.schema)
```

outputting

```
soma_joinid: int64 not null
int: int64 not null
str: large_string not null
int_cat: dictionary<values=int64, indices=int8, ordered=0> not null
byte: large_binary not null
str_cat: dictionary<values=string, indices=int8, ordered=0> not null
byte_cat: dictionary<values=string, indices=int8, ordered=0> not null
```

even though TileDB-Py reports correctly, with

```
#!/usr/bin/env python

import os
import sys
import tiledb

A = tiledb.open('test_dataframe')

for i in range(A.schema.nattr):
    attr = A.schema.attr(i)
    try:
        index_type = attr.dtype
        value_type = A.enum(attr.name).dtype
        print(f"enum name={attr.name} index_type={index_type.name} value_type={value_type.name}")
    except tiledb.cc.TileDBError:
        pass # not an enum attr
```

outputting

```
enum name=int_cat index_type=int8 value_type=int64
enum name=str_cat index_type=int8 value_type=str32
enum name=byte_cat index_type=int8 value_type=bytes
```

**Comment by nguyenv** [comment]
OK I will take a look tomorrow.

**Comment by johnkerl** [comment]
At present with R and this PR we have:

```
> sdf <- SOMADataFrameOpen('test_dataframe')
> sdf$schema()
Error: Unsupported data type: CHAR
```

so I'll add an R unit-test case here as well

**Comment by johnkerl** [comment]
Re my 

> so I'll add an R unit-test case here as well

we need first to fix R CI failures

**Comment by johnkerl** [comment]
All current CI is green. There are still a few things wrong I'm seeing interactively for which we had inadequate unit-test coverage. More commits coming soon.

**Comment by eddelbuettel** [comment]
Keep in mind that these arrow pieces will like benefit from / change under / get simpler the nanoarrow refactoring.  However I am currently blocked on that one as I see a lot of Python CI shrapnel (though only from IIRC three test files, the majority passes) that I cannot make real progress on ... as I am a complete neophyte when it comes to Python / C(++) interop.

**Comment by johnkerl** [comment]
Good commits on this PR; all green; further work tracked on #2324.

**Pull request merged**
